rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -129,254 +113,132 @@ impl SerialSubmitter {
-        let provider = self.mailbox.provider();
-        if !provider
-            .is_contract(&msg.message.recipient)
-            .await
-            .context(CTX)?
-        {
-            info!(
-                recipient=?msg.message.recipient,
-                "Could not fetch metadata: Recipient is not a contract"
-            );
-            return Ok(false);
-        }
-
-        let ism_address = self
-            .mailbox
-            .recipient_ism(msg.message.recipient)
-            .await
-            .context(CTX)?;
-
-        let Some(metadata) = self.metadata_builder
-            .build(ism_address, &msg.message)
-            .await?
-        else {
-            info!("Could not fetch metadata");
-            return Ok(false)
+            TxPrepareResult::CriticalFailure(e) => return Err(e),
         };
 
-        // Estimate transaction costs for the process call. If there are issues, it's
-        // likely that gas estimation has failed because the message is
-        // reverting. This is defined behavior, so we just log the error and
-        // move onto the next tick.
-        let tx_cost_estimate = match self
-            .mailbox
-            .process_estimate_costs(&msg.message, &metadata)
-            .await
-        {
-            Ok(tx_cost_estimate) => tx_cost_estimate,
-            Err(error) => {
-                info!(?error, "Error estimating process costs");
-                return Ok(false);
+        match op.submit().await {
+            TxRunResult::Success => {
+                self.metrics.txs_submitted.inc();
+                self.validation_queue.push(Reverse(op));
             }
-        };
-
-        // If the gas payment requirement hasn't been met, move to the next tick.
-        let Some(gas_limit) = self
-            .gas_payment_enforcer
-            .message_meets_gas_payment_requirement(&msg.message, &tx_cost_estimate)
-            .await?
-        else {
-            info!(?tx_cost_estimate, "Gas payment requirement not met yet");
-            return Ok(false);
-        };
-
-        // Go ahead and attempt processing of message to destination chain.
-        debug!(?gas_limit, "Ready to process message");
-
-        // TODO: consider differentiating types of processing errors, and pushing to the
-        //  front of the run queue for intermittent types of errors that can
-        //  occur even if a message's processing isn't reverting, e.g. timeouts
-        //  or txs being dropped from the mempool. To avoid consistently retrying
-        //  only these messages, the number of retries could be considered.
-
-        let gas_limit = tx_cost_estimate.gas_limit;
-
-        if let Some(max_limit) = self.transaction_gas_limit {
-            if gas_limit > max_limit {
-                info!("Message delivery estimated gas exceeds max gas limit");
-                return Ok(false);
+            TxRunResult::DoNotRetry => {
+                self.metrics.txs_submitted.inc();
```

Comment:
Why do we increment on DoNotRetry? Presumably we did not submit successfully, right?